File: app/src/main/cpp/PadEngine.cpp
#include "PadEngine.h"

#include <android/log.h>

#include <algorithm>
#include <cmath>

#include "OboeOutput.h"

#define LOG_TAG "PadEngine"
#define LOGW(...) __android_log_print(ANDROID_LOG_WARN, LOG_TAG, __VA_ARGS__)

namespace snipsnap {

PadEngine::PadEngine(int32_t preferredSampleRate) : preferredRate_(preferredSampleRate) {}

PadEngine::~PadEngine() {
    stop();
    // With the stream closed the audio thread is gone; every slot is ours.
    delete pending_.exchange(nullptr);
    delete retired_.exchange(nullptr);
    delete current_;
    current_ = nullptr;
}

bool PadEngine::start() {
    stop();
    restartNeeded_.store(false, std::memory_order_release);
    sharedMode_.store(false, std::memory_order_release);

    const OpenedOutput opened = openStereoFloatOutput(preferredRate_, this, this, stream_, LOG_TAG);
    if (!opened.ok) return false;
    sampleRate_ = opened.sampleRate;
    sharedMode_.store(opened.shared, std::memory_order_release);

    const oboe::Result started = stream_->requestStart();
    if (started != oboe::Result::OK) {
        LOGW("requestStart failed: %s", oboe::convertToText(started));
        stream_->close();
        stream_.reset();
        return false;
    }
    return true;
}

double PadEngine::latencyMillis() const {
    return latencyMillisOf(stream_);
}

void PadEngine::stop() {
    if (stream_) {
        stream_->stop();
        stream_->close();
        stream_.reset();
    }
    // The stream is closed, so the audio thread is gone and every slot -
    // the voices, both ends of both rings - is this thread's alone.
    //
    // The voices and the queued commands belonged to the stream that just
    // ended. `start()` calls this first, so without the sweep a route
    // change (headphones out, then the UI reopens) would pick those notes
    // up mid-sample seconds later, and the commands queued while there was
    // no stream would all fire at once into the new one. Every id is
    // reported ended, or the VoiceAllocator keeps it busy for ever.
    for (auto& v : voices_) {
        if (v.active) endVoice(v);
    }
    PadCommand dropped;
    while (commands_.pop(dropped)) {
        if (dropped.type == PadCommand::Type::NoteOn) ended_.push(dropped.voiceId);
    }
}

void PadEngine::onErrorAfterClose(oboe::AudioStream*, oboe::Result error) {
    LOGW("stream closed: %s", oboe::convertToText(error));
    restartNeeded_.store(true, std::memory_order_release);
}

// ---- the bank (UI thread) ---------------------------------------------------

void PadEngine::beginBank() {
    building_ = std::make_unique<Bank>();
}

int32_t PadEngine::addSample(std::vector<float>&& interleaved, int32_t channels, int32_t rate) {
    if (!building_) beginBank();
    BankSample s;
    s.frames = std::move(interleaved);
    s.channels = channels == 2 ? 2 : 1;
    s.rate = rate > 0 ? rate : 44100;
    building_->samples.push_back(std::move(s));
    return static_cast<int32_t>(building_->samples.size() - 1);
}

void PadEngine::commitBank() {
    if (!building_) return;
    // The generation moves before the bank is parked: from here every
    // command the UI pushes is for this bank, and the callback drops (and
    // reports) any it still finds stamped with the old one.
    building_->generation = uiGeneration_.load(std::memory_order_acquire) + 1;
    uiGeneration_.store(building_->generation, std::memory_order_release);
    // A bank the callback never got round to adopting is ours to free,
    // and so is the one it retired last time.
    delete pending_.exchange(building_.release(), std::memory_order_acq_rel);
    delete retired_.exchange(nullptr, std::memory_order_acq_rel);
}

size_t PadEngine::drainEnded(int32_t* out, size_t max) {
    size_t n = 0;
    int32_t id;
    while (n < max && ended_.pop(id)) out[n++] = id;
    return n;
}

// ---- audio thread from here down ---------------------------------------------

void PadEngine::adoptPendingBank() {
    if (pending_.load(std::memory_order_acquire) == nullptr) return;
    // The UI has not collected the last retiree yet: wait for the next
    // callback rather than hand anything back. retired_ only goes
    // null -> non-null here, so the check-then-store cannot race.
    if (retired_.load(std::memory_order_acquire) != nullptr) return;
    Bank* incoming = pending_.exchange(nullptr, std::memory_order_acq_rel);
    if (!incoming) return;
    // Every voice belonged to the old bank: silence them and say so.
    for (auto& v : voices_) {
        if (v.active) endVoice(v);
    }
    retired_.store(current_, std::memory_order_release);
    current_ = incoming;
}

void PadEngine::endVoice(Voice& v) {
    v.active = false;
    v.fadeStep = 0.0f;
    // A full ring drops the report; the UI drains every frame, so that
    // takes hundreds of endings inside one screen frame.
    ended_.push(v.id);
}

PadEngine::Voice& PadEngine::freeVoice() {
    // An idle slot first; then the voice already fading out furthest along;
    // then the oldest - the ear misses old tails least.
    Voice* pick = nullptr;
    for (auto& v : voices_) {
        if (!v.active) return v;
    }
    for (auto& v : voices_) {
        if (v.fadeStep > 0.0f && (!pick || v.fade < pick->fade)) pick = &v;
    }
    if (!pick) {
        for (auto& v : voices_) {
            if (!pick || v.serial < pick->serial) pick = &v;
        }
    }
    endVoice(*pick);
    return *pick;
}

void PadEngine::apply(const PadCommand& c) {
    // A command for another bank (queued across a kit swap) is honest
    // silence: a NoteOn is reported ended so the allocator lets it go, a
    // Stop is moot (the swap silenced every voice), AllOff always applies.
    const bool stale = c.type != PadCommand::Type::AllOff &&
        (!current_ || c.generation != current_->generation);
    if (stale) {
        if (c.type == PadCommand::Type::NoteOn) ended_.push(c.voiceId);
        return;
    }
    switch (c.type) {
        case PadCommand::Type::NoteOn: {
            if (!current_ || c.sample < 0 || c.sample >= static_cast<int32_t>(current_->samples.size())) {
                ended_.push(c.voiceId);  // nothing to play: the allocator must not count it
                return;
            }
            const BankSample& s = current_->samples[static_cast<size_t>(c.sample)];
            const int64_t frames = s.frameCount();
            const int64_t start = std::max<int64_t>(0, std::min<int64_t>(c.start, frames));
            const int64_t end = std::max<int64_t>(start, std::min<int64_t>(c.end, frames));
            if (end - start < 2) {
                ended_.push(c.voiceId);
                return;
            }
            // The speed and the gains cross JNI as raw numbers, and render()
            // only ever guards the far end of the read. A speed of zero
            // freezes the read on one frame for ever - the voice can never
            // reach `end`, so nothing frees it but a steal - and a negative
            // one walks `pos` back off the front of the buffer. Neither is a
            // note, so refuse it the way a bad index is refused; the top is
            // clamped rather than refused, because an absurd tune is still a
            // note somebody asked for.
            const double speed = (static_cast<double>(s.rate) / static_cast<double>(sampleRate_)) * c.pitch;
            if (!std::isfinite(speed) || speed <= 0.0) {
                ended_.push(c.voiceId);
                return;
            }
            Voice& v = freeVoice();
            v.active = true;
            v.id = c.voiceId;
            v.sample = c.sample;
            v.start = start;
            v.end = end;
            v.reverse = c.reverse;
            // Backwards starts at the last frame of the window and walks down.
            v.pos = c.reverse ? static_cast<double>(end - 1) : static_cast<double>(start);
            // A loop wraps from the last frame back to loopStart; a loop that
            // would be empty (start at or past end - 1) plays once instead.
            v.loopStart = (c.loopStart >= 0 && c.loopStart < end - 1) ? c.loopStart : -1;
            v.inc = std::min(speed, kMaxSpeed);
            // A gain that is not a number would make the whole mix one: the
            // clamp at the end of render() misses NaN, because both of its
            // comparisons are false.
            v.gainL = std::isfinite(c.gainL) ? c.gainL : 0.0f;
            v.gainR = std::isfinite(c.gainR) ? c.gainR : 0.0f;
            v.targetL = v.gainL;
            v.targetR = v.gainR;
            v.gainRamp = 0;  // a new note starts where it was told, no glide
            v.fade = 1.0f;
            v.fadeStep = 0.0f;
            v.serial = ++serial_;
            return;
        }
        case PadCommand::Type::SetGain: {
            // A fader move on a voice already sounding. Gliding rather than
            // jumping, because a drag sends one of these per screen frame
            // and a step per frame is a zipper.
            const float glide = std::max(1.0f, c.fadeMs * 0.001f * static_cast<float>(sampleRate_));
            for (auto& v : voices_) {
                if (!v.active || v.id != c.voiceId) continue;
                // A gain that is not a number leaves the fader where it was.
                v.targetL = std::isfinite(c.gainL) ? c.gainL : v.gainL;
                v.targetR = std::isfinite(c.gainR) ? c.gainR : v.gainR;
                v.gainRamp = static_cast<int32_t>(glide);
            }
            return;
        }
        case PadCommand::Type::Stop:
        case PadCommand::Type::AllOff: {
            const float fadeFrames = std::max(1.0f, c.fadeMs * 0.001f * static_cast<float>(sampleRate_));
            for (auto& v : voices_) {
                if (!v.active) continue;
                if (c.type == PadCommand::Type::Stop && v.id != c.voiceId) continue;
                if (v.fadeStep <= 0.0f) v.fadeStep = 1.0f / fadeFrames;
            }
            return;
        }
    }
}

void PadEngine::render(float* out, int32_t numFrames) {
    for (int32_t i = 0; i < numFrames * 2; ++i) out[i] = 0.0f;
    if (!current_) return;
    for (auto& v : voices_) {
        if (!v.active) continue;
        const BankSample& s = current_->samples[static_cast<size_t>(v.sample)];
        const float* f = s.frames.data();
        const int64_t last = v.end - 1;
        for (int32_t i = 0; i < numFrames; ++i) {
            // A fast voice over a short loop can cross the far end more
            // than once in one frame: wrap until the read is back inside.
            // Backwards the loop runs the other way - off the bottom at
            // loopStart, back up to the last frame - and the window ends
            // at `start` rather than at `end`.
            if (v.reverse) {
                while (v.loopStart >= 0 && v.pos <= static_cast<double>(v.loopStart)) {
                    v.pos += static_cast<double>(last - v.loopStart);
                }
                if (v.pos < static_cast<double>(v.start)) {
                    endVoice(v);
                    break;
                }
            } else {
                while (v.loopStart >= 0 && v.pos >= static_cast<double>(last)) {
                    v.pos -= static_cast<double>(last - v.loopStart);
                }
            }
            const int64_t i0 = static_cast<int64_t>(v.pos);
            if (i0 >= v.end) {
                endVoice(v);
                break;
            }
            const int64_t i1 = std::min<int64_t>(i0 + 1, last);
            const float frac = static_cast<float>(v.pos - static_cast<double>(i0));
            float l, r;
            if (s.channels == 2) {
                const float l0 = f[2 * i0], l1 = f[2 * i1];
                const float r0 = f[2 * i0 + 1], r1 = f[2 * i1 + 1];
                l = l0 + (l1 - l0) * frac;
                r = r0 + (r1 - r0) * frac;
            } else {
                const float m = f[i0] + (f[i1] - f[i0]) * frac;
                l = r = m;
            }
            // The glide: an exact linear arrival, because closing 1/n of
            // what is left with n falling by one each sample *is* a line.
            if (v.gainRamp > 0) {
                const float step = 1.0f / static_cast<float>(v.gainRamp);
                v.gainL += (v.targetL - v.gainL) * step;
                v.gainR += (v.targetR - v.gainR) * step;
                --v.gainRamp;
            }
            if (v.fadeStep > 0.0f) {
                v.fade -= v.fadeStep;
                if (v.fade <= 0.0f) {
                    endVoice(v);
                    break;
                }
            }
            out[2 * i] += l * v.gainL * v.fade;
            out[2 * i + 1] += r * v.gainR * v.fade;
            // The interpolation above is positional, not directional: the
            // pair either side of `pos` is the same pair whichever way the
            // read is travelling. Only the step changes sign.
            v.pos += v.reverse ? -v.inc : v.inc;
        }
    }
    // Thirty-two voices at kit levels rarely sum past full scale; when they
    // do, a clamp beats wrapping.
    for (int32_t i = 0; i < numFrames * 2; ++i) {
        if (out[i] > 1.0f) out[i] = 1.0f;
        else if (out[i] < -1.0f) out[i] = -1.0f;
    }
}

oboe::DataCallbackResult PadEngine::onAudioReady(oboe::AudioStream*, void* audioData, int32_t numFrames) {
    adoptPendingBank();
    PadCommand c;
    while (commands_.pop(c)) apply(c);
    auto* out = static_cast<float*>(audioData);
    render(out, numFrames);
    // The print tap takes the bus after the mix, so what it captures is
    // what the speaker got: the clamp, the stolen voices and the fades
    // are already in these frames. It is a no-op unless a print is armed.
    print_.record(out, static_cast<size_t>(numFrames));
    return oboe::DataCallbackResult::Continue;
}

}  // namespace snipsnap

File: app/src/main/kotlin/com/snipsnap/app/PadEngine.kt
package com.snipsnap.app

import com.snipsnap.app.ui.TAPE_LOAD_MAX_SEC
import com.snipsnap.audio.DrumSynth
import com.snipsnap.audio.Snip
import com.snipsnap.audio.WavReader
import com.snipsnap.kit.KitPad
import com.snipsnap.shell.PadHit
import java.io.File

/**
 * M4's pad voice: the Kotlin owner of one native [NativePads] engine,
 * replacing the interim SoundPool player for PLAY. A kit's samples are
 * read once ([load], off the main thread - it reads WAVs) into a bank
 * the callback adopts whole; a hit resolves through `PadHit` (`:shell`,
 * tested: the layer, the chain slice, level, pan, tune) and crosses the
 * bridge as one command keyed by the allocator's voice id; the engine
 * reports each voice's end back through [drainEnded], so the allocator
 * learns of an ending when it happens.
 *
 * Any thread but the audio thread; calls are serialised. [close] is
 * idempotent and every call after it is a no-op.
 */
class PadEngine(preferredSampleRate: Int) {

    private var handle: Long = NativePads.create(preferredSampleRate)
    private val open get() = handle != 0L

    /** True between a successful [start] and [close]. See [isUp] for whether a callback is actually running. */
    @Volatile
    var running: Boolean = false
        private set

    /**
     * A callback is running right now: started, not closed, and the native
     * side has not flagged a dead stream. The flag flips on the audio
     * thread's own error callback, so this is the truth even in the window
     * before the screen's frame loop gets round to the restart - a hit in
     * that window is refused rather than counted by nothing.
     */
    @Synchronized
    fun isUp(): Boolean = open && running && !NativePads.needsRestart(handle)

    /**
     * One bank build at a time. The per-call `synchronized(this)` below
     * keeps each native call safe on its own, but a *build* is a sequence —
     * begin, add, add, …, commit — and the native builder is one slot.
     * Two overlapping builds would have the second `beginBank` throw away
     * the first's half-filled bank, leaving its sample indices pointing
     * into the other's, and whichever committed last would win with a
     * mixture of two kits.
     *
     * Deliberately *not* `this`: [close] and the command calls synchronize
     * on the instance, and holding those across megabytes of copying is
     * what makes a close on the main thread wait. This lock serialises
     * builders against each other and nothing else.
     */
    private val bankLock = Any()

    private var sampleIndex: Map<String, Int> = emptyMap()
    private var framesOf: Map<String, Long> = emptyMap()
    private val hits = HashMap<Int, Int>()

    /**
     * The count-in click's reserved bank index, valid only after [load];
     * -1 before then or if the engine has no stream. This is a native bank
     * index — the same numbering [NativePads.addSample] hands back for
     * every kit sample — not a [KitPad.slot] (1..128); the two are
     * different spaces and never compared. [load] appends the click as
     * the last two samples of the bank it rebuilds every time (accent,
     * then this one, so this index is always the accent's plus one), and
     * neither can ever collide with a kit sample's index: every kit file
     * is indexed first, in file order, before either click is added, so
     * this index is always past every one of them.
     */
    var clickSampleIndex: Int = -1
        private set

    /** The click's accented (downbeat) sibling — one bank slot before [clickSampleIndex], same reasoning. */
    private var clickAccentIndex: Int = -1
    private var clickFrames: Long = 0L
    private var clickAccentFrames: Long = 0L

    @Synchronized
    fun start(): Boolean {
        running = open && NativePads.start(handle)
        return running
    }

    @Synchronized
    fun needsRestart(): Boolean = open && NativePads.needsRestart(handle)

    /** The device refused an exclusive stream; the shared fallback is playing. */
    @Synchronized
    fun isShared(): Boolean = open && NativePads.isShared(handle)

    /**
     * The stream's round-trip latency in ms, or null when there is none or
     * the device will not say. Cheap but not free (it reads a timestamp),
     * so poll it about once a second, not per frame.
     */
    @Synchronized
    fun latencyMillis(): Double? =
        if (open && running) NativePads.latencyMillis(handle).takeIf { it > 0.0 } else null

    /**
     * Read every sample the kit's pads and layers name and hand the bank to
     * the engine. Blocking disk IO: call it from an IO dispatcher. A WAV
     * that will not read is left out, and its pad plays nothing (a hit
     * resolves to null) rather than something else.
     */
    fun load(entry: KitShelf.Entry) = synchronized(bankLock) { loadLocked(entry) }

    private fun loadLocked(entry: KitShelf.Entry) {
        val files = LinkedHashSet<String>()
        for (pad in entry.kit.pads) {
            files += pad.sampleFile
            pad.velocityLayers.forEach { files += it.sampleFile }
        }
        val index = HashMap<String, Int>()
        val frames = HashMap<String, Long>()
        var engineRate = 44_100
        synchronized(this) {
            if (!open) return
            NativePads.beginBank(handle)
            engineRate = NativePads.sampleRate(handle)
        }
        for (file in files) {
            // `file` is a kit pad sample, produced only by KitBuilderModel.assign
            // from an already-bounded Snip — readCapped's 600s ceiling is defense
            // in depth, not expected to ever bind.
            val snip = runCatching { WavReader.readCapped(File(entry.dir, file), TAPE_LOAD_MAX_SEC).snip }.getOrNull() ?: continue
            val i = synchronized(this) {
                if (!open) return
                NativePads.addSample(handle, snip.samples, snip.channels, snip.sampleRate)
            }
            index[file] = i
            frames[file] = snip.frameCount.toLong()
        }
        // The count-in click, synthesized (never a bundled asset) and
        // appended last, inside this same bank build — not a second
        // beginBank/commitBank pair, and never loadSnips, which replaces
        // the whole bank and would silence every pad just added above.
        val accentClick = DrumSynth.click(engineRate, accent = true)
        val regularClick = DrumSynth.click(engineRate, accent = false)
        val accentIdx = synchronized(this) {
            if (!open) return
            NativePads.addSample(handle, accentClick.samples, accentClick.channels, accentClick.sampleRate)
        }
        val regularIdx = synchronized(this) {
            if (!open) return
            NativePads.addSample(handle, regularClick.samples, regularClick.channels, regularClick.sampleRate)
        }
        synchronized(this) {
            if (!open) return
            NativePads.commitBank(handle)
            sampleIndex = index
            framesOf = frames
            hits.clear()
            clickAccentIndex = accentIdx
            clickSampleIndex = regularIdx
            clickAccentFrames = accentClick.frameCount.toLong()
            clickFrames = regularClick.frameCount.toLong()
        }
    }

    /**
     * Bank audio already in hand rather than files on the shelf — what
     * SPLIT has once a separation has run. Returns each snip's bank index
     * in the order given, or an empty list when there is no engine.
     *
     * A bank holds one thing at a time, so the kit index [load] built is
     * cleared with it: after this the engine plays these snips and nothing
     * else. Copies every sample across the bridge, so keep it off the main
     * thread.
     *
     * Locked per native call rather than for the whole run, exactly as
     * [load] is: a split's three buffers are megabytes, and holding the
     * monitor across all of them would make a [close] on the main thread —
     * leaving the screen mid-load — wait for every copy to finish.
     */
    fun loadSnips(snips: List<Snip>): List<Int> = synchronized(bankLock) { loadSnipsLocked(snips) }

    private fun loadSnipsLocked(snips: List<Snip>): List<Int> {
        synchronized(this) {
            if (!open) return emptyList()
            NativePads.beginBank(handle)
        }
        val indices = ArrayList<Int>(snips.size)
        for (snip in snips) {
            indices += synchronized(this) {
                if (!open) return emptyList()
                NativePads.addSample(handle, snip.samples, snip.channels, snip.sampleRate)
            }
        }
        synchronized(this) {
            if (!open) return emptyList()
            NativePads.commitBank(handle)
            sampleIndex = emptyMap()
            framesOf = emptyMap()
            hits.clear()
            // This bank replaces the one `load` built (see the KDoc above) -
            // any click indices it recorded now point into a bank that no
            // longer exists. Clearing them keeps `clickHit`'s -1 guard
            // truthful instead of leaving a third, undocumented state where
            // it addresses a stale or wrong sample.
            clickSampleIndex = -1
            clickAccentIndex = -1
            clickFrames = 0L
            clickAccentFrames = 0L
        }
        return indices
    }

    /** The loaded sample's length in frames, or null when it never loaded. */
    @Synchronized
    fun frames(sampleFile: String): Long? = framesOf[sampleFile]

    /**
     * Play [pad] at [velocity] as voice [voiceId]. False when no stream is
     * running, the pad has nothing loaded to play, or the command ring was
     * full - so the caller can tell its allocator the voice never sounded
     * and nothing counts a voice that no callback will ever end.
     */
    @Synchronized
    fun hit(pad: KitPad, velocity: Float, voiceId: Int): Boolean {
        if (!isUp()) return false
        val n = hits[pad.slot] ?: 0
        hits[pad.slot] = n + 1
        val hit = PadHit.resolve(pad, velocity, n) { framesOf[it] } ?: return false
        val sample = sampleIndex[hit.sampleFile] ?: return false
        return NativePads.noteOn(
            handle, voiceId, sample,
            hit.startFrame, hit.endFrameExclusive, -1L, hit.gainLeft, hit.gainRight, hit.pitchRatio,
            reverse = false,
        )
    }

    /**
     * Start [layers] of one loaded sample together. Every layer reads the
     * same window at the same speed - they are one sound taken apart - so
     * only the gains and the direction differ, and they are published to
     * the callback as one group: three layers can never start a buffer
     * apart, which would be heard as a flam and chased for hours.
     *
     * False means *nothing* was queued: the caller still owns every id it
     * was given, and none of them is sounding.
     */
    @Synchronized
    fun hitLayers(layers: List<Layer>, startFrame: Long, endFrame: Long, loopStart: Long, pitch: Double): Boolean {
        if (!isUp() || layers.isEmpty()) return false
        // The native side refuses a group larger than the engine has voices;
        // saying so here means four arrays are never built for it.
        if (layers.size > MAX_VOICES) return false
        return NativePads.noteOnLayers(
            handle,
            IntArray(layers.size) { layers[it].voiceId },
            IntArray(layers.size) { layers[it].sample },
            startFrame, endFrame, loopStart,
            FloatArray(layers.size) { layers[it].gainLeft },
            FloatArray(layers.size) { layers[it].gainRight },
            pitch,
            BooleanArray(layers.size) { layers[it].reverse },
        )
    }

    /** One voice of a group: which bank sample, at what level, which way round. */
    data class Layer(
        val voiceId: Int,
        val sample: Int,
        val gainLeft: Float,
        val gainRight: Float,
        val reverse: Boolean,
    )

    /**
     * Play the count-in click at [voiceId], one-shot, full length —
     * [accent] for the downbeat, false for the other three beats. Goes
     * through [hitLayers] with a single raw bank index, exactly the
     * `SplitScreen` precedent for sounding a bank sample with no
     * [KitPad] involved: this is a timing cue, not a kit sound, so it
     * never touches a choke group, a pad slot, or `VoiceAllocator` — the
     * voice id is the caller's to give and own, same as every other
     * `hitLayers` call.
     *
     * False exactly when [hit] would be false: no stream running, or
     * (here) the click never loaded — [load] hasn't been called yet, or
     * the bank build never got as far as adding it.
     *
     * The caller owns [voiceId] and must supply one outside any
     * `VoiceAllocator`'s range on the same engine (its own ids start at 1
     * and only grow, so any id `<= 0` is safe forever): the native side
     * stamps whatever id it is given onto a voice with no dedup against
     * ids already sounding, so an id the allocator could also hand out
     * would let a click's own ending get reported as a pad's, or vice
     * versa, to whichever side is listening on [drainEnded].
     */
    @Synchronized
    fun clickHit(voiceId: Int, accent: Boolean): Boolean {
        val sample = if (accent) clickAccentIndex else clickSampleIndex
        val frames = if (accent) clickAccentFrames else clickFrames
        if (sample < 0 || frames <= 0L) return false
        return hitLayers(
            layers = listOf(Layer(voiceId = voiceId, sample = sample, gainLeft = CLICK_GAIN, gainRight = CLICK_GAIN, reverse = false)),
            startFrame = 0L,
            endFrame = frames,
            loopStart = -1L,
            pitch = 1.0,
        )
    }

    /**
     * Move a sounding voice's gains, gliding over [glideMs] rather than
     * stepping. This is what a fader is: retriggering the note on every
     * drag frame would be a click per frame, and a step per frame a
     * zipper. A voice that has already ended simply ignores it.
     */
    @Synchronized
    fun setGain(voiceId: Int, gainLeft: Float, gainRight: Float, glideMs: Float = FADER_GLIDE_MS) {
        if (open) NativePads.setVoiceGain(handle, voiceId, gainLeft, gainRight, glideMs)
    }

    /** Stop one voice with a short fade: a choke, a steal, a gate's release. */
    @Synchronized
    fun stop(voiceId: Int, fadeMs: Float = CHOKE_FADE_MS) {
        if (open) NativePads.stopVoice(handle, voiceId, fadeMs)
    }

    /** Everything, with a slightly longer fade (leaving PLAY, the panic). */
    @Synchronized
    fun allOff(fadeMs: Float = PANIC_FADE_MS) {
        if (open) NativePads.allOff(handle, fadeMs)
    }

    /** The ids of voices that ended since the last call; drain at screen rate. */
    @Synchronized
    fun drainEnded(): IntArray = if (open) NativePads.drainEnded(handle) else IntArray(0)

    @Synchronized
    fun close() {
        if (!open) return
        running = false
        NativePads.stop(handle)
        NativePads.destroy(handle)
        handle = 0L
    }

    companion object {
        /** The engine's polyphony; the VoiceAllocator in PLAY is built at the same number. */
        const val MAX_VOICES = 32

        /**
         * A fader's glide. Long enough that a drag is smooth, short enough
         * that letting go feels immediate - about two screen frames.
         */
        const val FADER_GLIDE_MS = 30f

        /** A choke is a short fade, not a cut: long enough to spare the click, short enough to read as a cut. */
        const val CHOKE_FADE_MS = 5f
        const val PANIC_FADE_MS = 20f

        /** The count-in click's fixed level — a cue meant to sit under the kit, not a mixed-in sound. */
        const val CLICK_GAIN = 0.6f
    }
}
